### ISSUE TITLE
测试检索时的问题

### ISSUE THREAD

**Issue by fanxuansoft** [flyingpot/pytorch_deephash#10]
您好，借用您的代码，将网络模型换为了VGG16，训练之后进行mAP测试时，直接用：python mAP.py --pretrained 测试的话，程序返回给我的mAP达到了91%.
但是我无法确认这个结果的准确性，我对训练集图片进行了标识，然后sort_indices数组的前20位出来当做索引，但是这20个图片所对应的图片都不是我所检索的图片。
想请教一下，如果我想要返回单次检索的top10，应该如何做？或者说mAP返回的准确率是基于top几的数据来计算的。
期待您的回复。

**Comment by spoonnotfound** [comment]
我这个代码的mAP是基于top-k(k=训练集的大小)的，你可以自己调整。然后你要top-10的话就对sort_indices做[:10]切片就行，然后用trn_label[sort_indices]得到对应的label。你说对应的图片不是检索的图片，你确定吗？这个mAP的计算我是完全按照原作者的[matlab代码](https://github.com/kevinlin311tw/caffe-cvprw15/blob/master/analysis/precision.m)修改的。

**Comment by fanxuansoft** [comment]
依照您的回复，修改如下：`AP = np.zeros(query_times)
    Ns = np.arange(1, 10 + 1)
    total_time_start = time.time()
    for i in range(query_times):
        print('Query ', i+1)
        query_label = tst_label[i]
        query_binary = tst_binary[i,:]
        query_result = np.count_nonzero(query_binary != trn_binary, axis=1)    #don't need to divide binary length
        sort_indices = np.argsort(query_result)
        # np.set_printoptions(edgeitems=20)
        sort_indices = sort_indices[:10]
        print(sort_indices)
        buffer_yes= np.equal(query_label, trn_label[sort_indices]).astype(int)
        print(buffer_yes)
        P = np.cumsum(buffer_yes) / Ns
        AP[i] = np.sum(P * buffer_yes) /sum(buffer_yes)
        print(AP[i])
    map = np.mean(AP)
    print(map)
    print('total query time = ', time.time() - total_time_start)`

这部分代码改的应该没有问题吧，测试结果（贴的是最后一张图片的测试结果）：
Query  6330
[21130 29311 46251 54235 11228 31650 61944 25820 41582 51890]
[1 1 1 1 1 1 1 1 1 1]
1.0
nan
total query time =  38.333385705947876

测试数据6330对应的图片为名：金燕子00091
sort_indices前10位对应的图片为（前面为图片名，后面为索引号）：
2A3210 (3)_0_5358 21130
3M3086 (2)_0_6238 29311
9511F-156 (1)_0_9386 46251
K3448 (2)_0_5108 54235
1C8317 (2)_0_68 11228
8027 (5)_0_7877 31650
sm195 (1)_0_5247 61944
3101 (2)_0_5389 25820
9153-887 (1)_0_1141 41582
K3247 (1)_0_1128 51890

如果您确认mAP没有问题的话，那么有可能是我给图片标识的有问题。
想问您一下，sort_indices中的数字对应的是什么？是否是图片的文件夹中的位置还是什么？

**Comment by fanxuansoft** [comment]
上面的代码贴的不好，precision函数：
def precision(trn_binary, trn_label, tst_binary, tst_label):
    trn_binary = trn_binary.cpu().numpy()
    trn_binary = np.asarray(trn_binary, np.int32)
    trn_label = trn_label.cpu().numpy()
    tst_binary = tst_binary.cpu().numpy()
    tst_binary = np.asarray(tst_binary, np.int32)
    tst_label = tst_label.cpu().numpy()
    query_times = tst_binary.shape[0]
    trainset_len = train_binary.shape[0]
    AP = np.zeros(query_times)
    Ns = np.arange(1, 10 + 1)
    total_time_start = time.time()
    for i in range(query_times):
        print('Query ', i+1)
        query_label = tst_label[i]
        query_binary = tst_binary[i,:]
        query_result = np.count_nonzero(query_binary != trn_binary, axis=1)    #don't need to divide binary length
        sort_indices = np.argsort(query_result)
        # np.set_printoptions(edgeitems=20)
        sort_indices = sort_indices[:10]
        print(sort_indices)
        buffer_yes= np.equal(query_label, trn_label[sort_indices]).astype(int)
        print(buffer_yes)
        P = np.cumsum(buffer_yes) / Ns
        AP[i] = np.sum(P * buffer_yes) /sum(buffer_yes)
        print(AP[i])
    map = np.mean(AP)
    print(map)
    print('total query time = ', time.time() - total_time_start)

**Comment by spoonnotfound** [comment]
sort_indices数字对应的的是汉明距离从小到大对应的索引值。我不是很清楚怎么映射会图片，你可以自己研究一下。但是检索是肯定没问题的。

**Comment by fanxuansoft** [comment]
好的，谢谢

**Comment by fanxuansoft** [comment]
输出为图片我已经调试出来了，很简单，加载图片的时候DataLoader shuffle设为False，这时候检索时sort_indices返回的值就是训练集图片的索引（从0开始），顺便验证了一下，这个检索是没有问题的，感谢

**Comment by world2025** [comment]
@fanxuansoft ,你好，你是在训练集和测试集都将shuffle设置为false吗？还有请问你有计算recall吗？谢谢

**Comment by world2025** [comment]
@fanxuansoft 还请问你就是如何可视化结果的，能分享下代码吗？谢谢

**Comment by InitAction** [comment]
@fanxuansoft 你好，能分享下vgg16模型的代码吗？

**Comment by InitAction** [comment]
@fanxuansoft 你好，能分享下vgg16模型的代码吗？
